File: src/main/java/com/felipe/booking/domain/usecase/CreateBookingUseCase.java
package com.felipe.booking.domain.usecase;

import com.felipe.booking.domain.gateway.BookingDataSourceGateway;
import com.felipe.booking.domain.model.Booking;
import com.felipe.booking.domain.model.BookingLock;
import org.springframework.beans.factory.annotation.Autowired;
import reactor.core.publisher.Mono;

public class CreateBookingUseCase {

    @Autowired
    private ValidateBookingUseCase validateBookingUseCase;

    @Autowired
    private BookingDataSourceGateway bookingDataSourceGateway;


    public Mono<Booking> execute(Booking booking) {
        return bookingDataSourceGateway
                .getLock()
                .flatMap (lock -> saveBooking(booking, lock));
    }

    private Mono<Booking> saveBooking(Booking booking, BookingLock lock) {
        return validateBookingUseCase
                .validate(booking)
                .flatMap (it -> bookingDataSourceGateway.createBooking(booking, lock));
        }
}

File: src/main/kotlin/com/felipe/booking/alten/domain/usecase/CreateBookingUseCase.kt
package com.felipe.booking.alten.domain.usecase

import com.felipe.booking.alten.domain.gateway.BookingDataSourceGateway
import com.felipe.booking.alten.domain.model.Booking
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono

@Component
class CreateBookingUseCase(
    private val bookingRepositoryGateway: BookingDataSourceGateway
) {

    fun execute(): Mono<Booking> =
        bookingRepositoryGateway
            .getLock()
            .flatMap { bookingRepositoryGateway.createBooking(it) }
}